### PR TITLE
docs: add documentation

### PR DESCRIPTION
**Pull request by kassianh**

### CODE REVIEW

File: README.tfdoc.hcl [hunk 1]
```diff
@@ -0,0 +1,1223 @@
...
+          type        = object(forward)
+          description = <<-END
+            Configuration block for creating an action that distributes requests
+            among one or more target groups. Specify only if `type` is `forward`.
+            If you specify both `forward` block and `target_group_arn`
+            attribute, you can specify only one target group using `forward` and
+            it must be the same target group specified in `target_group_arn`.
+          END
+
+          attribute "target_group" {
```
Comment (soerenmartius) [inline README.tfdoc.hcl:396]:
```suggestion
          attribute "target_groups" {
```

File: README.tfdoc.hcl [hunk 2]
```diff
@@ -0,0 +1,1223 @@
...
+            type        = set(target_goup)
```
Comment (soerenmartius) [inline README.tfdoc.hcl:397]:
```suggestion
            type        = set(target_group)
```

File: README.tfdoc.hcl [hunk 3]
```diff
@@ -0,0 +1,1223 @@
...
+            default     = "HTTP_302"
+            description = <<-END
+              HTTP redirect code. The redirect is either permanent (`HTTP_301`)
+              or temporary (`HTTP_302`).
+            END
+          }
+
+          attribute "host" {
+            type        = string
+            default     = "HTTP_302"
```
Comment (soerenmartius) [inline README.tfdoc.hcl:462]:
```suggestion
```

File: README.tfdoc.hcl [hunk 4]
```diff
@@ -0,0 +1,1223 @@
...
+    section {
+      title = "Extended Resource Configuration"
+
+      variable "rules" {
+        type        = list(rule)
+        default     = []
+        description = <<-END
+          A list of rules to be attached to the created listener.
+        END
+
```
Comment (soerenmartius) [inline README.tfdoc.hcl:524]: ID attribute seems to be missing

Comment (soerenmartius) [reply README.tfdoc.hcl:524]: [**https://github.com/mineiros-io/terraform-aws-lb-listener/blob/main/variables.tf#L181**](https://github.com/mineiros-io/terraform-aws-lb-listener/blob/main/variables.tf#L181)

File: README.tfdoc.hcl [hunk 5]
```diff
@@ -0,0 +1,1223 @@
...
+            type        = object(forward)
+            description = <<-END
+              Configuration block for creating an action that distributes requests
+              among one or more target groups. Specify only if `type` is `forward`.
+              If you specify both `forward` block and `target_group_arn`
+              attribute, you can specify only one target group using `forward` and
+              it must be the same target group specified in `target_group_arn`.
+            END
+
+            attribute "target_group" {
```
Comment (soerenmartius) [inline README.tfdoc.hcl:609]:
```suggestion
            attribute "target_groups" {
```

File: README.tfdoc.hcl [hunk 6]
```diff
@@ -0,0 +1,1223 @@
...
+              default     = "HTTP_302"
+              description = <<-END
+                HTTP redirect code. The redirect is either permanent (`HTTP_301`)
+                or temporary (`HTTP_302`).
+              END
+            }
+
+            attribute "host" {
+              type        = string
+              default     = "HTTP_302"
```
Comment (soerenmartius) [inline README.tfdoc.hcl:675]:
```suggestion
```

File: README.tfdoc.hcl [hunk 7]
```diff
@@ -0,0 +1,1223 @@
...
+          attribute "host_header" {
+            type        = object(host_header)
+            default     = []
+            description = <<-END
+              The maximum size of each pattern is 128 characters. Comparison is
+              case insensitive. Wildcard characters supported: `*` (matches 0 or
+              more characters) and `?` (matches exactly 1 character). Only one
+              pattern needs to match for the condition to be satisfied.
+            END
+
```
Comment (soerenmartius) [inline README.tfdoc.hcl:736]: `http_header_name` seems to be missing https://github.com/mineiros-io/terraform-aws-lb-listener/blob/main/variables.tf#L251

File: README.tfdoc.hcl [hunk 8]
```diff
@@ -0,0 +1,1223 @@
...
+              Domain prefix or fully-qualified domain name of the Cognito user
+              pool.
+            END
+          }
+
+          attribute "authentication_request_extra_params" {
+            type        = map(string)
+            description = <<-END
+              Query parameters to include in the redirect request to the
+              authorization endpoint. Max: 10.
```
Comment (soerenmartius) [inline README.tfdoc.hcl:870]:
```suggestion
              authorization endpoint. Accepts a maximum of 10 extra parameters.
```

**Review by soerenmartius — DISMISSED** [review]

File: README.tfdoc.hcl [hunk 9]
```diff
@@ -0,0 +1,1265 @@
...
+          description = <<-END
+            Order for the action. This value is required for rules with multiple
+            actions. The action with the lowest value for order is performed
+            first. Valid values are between `1` and `50000`.
+          END
+        }
+
+        attribute "authenticate_cognito" {
+          type        = object(authenticate_cognito)
+          default     = []
```
Comment (soerenmartius) [inline README.tfdoc.hcl:191]:
```suggestion
```

File: README.tfdoc.hcl [hunk 10]
```diff
@@ -0,0 +1,1265 @@
...
+          description = <<-END
+            Configuration block for using Amazon Cognito to authenticate users.
+            Specify only when `type` is `authenticate-cognito`.
```
Comment (soerenmartius) [inline README.tfdoc.hcl:194]:
```suggestion
```

This is not required as we abstract away the complexity (type isn't required in this case)

File: README.tfdoc.hcl [hunk 11]
```diff
@@ -0,0 +1,1265 @@
...
+        END
+
+        attribute "type" {
+          required    = true
+          type        = string
+          description = <<-END
+            Type of routing action. Valid values are `forward`, `redirect`,
+            `fixed-response`, `authenticate-cognito` and `authenticate-oidc`.
+          END
+        }
```
Comment (soerenmartius) [inline README.tfdoc.hcl:169]:
```suggestion
```

Can be removed, we abstracted the `type` attribute

File: README.tfdoc.hcl (hunk 10 above)
Comment (soerenmartius) [reply README.tfdoc.hcl:194]: for example: https://github.com/mineiros-io/terraform-aws-lb-listener/blob/main/main.tf#L41

File: README.tfdoc.hcl [hunk 12]
```diff
@@ -0,0 +1,1265 @@
...
+            required    = true
+            type        = string
+            description = <<-END
+              Domain prefix or fully-qualified domain name of the Cognito user
+              pool.
+            END
+          }
+
+          attribute "authentication_request_extra_params" {
+            type        = map(string)
```
Comment (soerenmartius) [inline README.tfdoc.hcl:223]:
```suggestion
            type        = list(authentication_request_extra_param)
```

This is a list of objects afaics

Comment (soerenmartius) [reply README.tfdoc.hcl:223]: https://github.com/mineiros-io/terraform-aws-lb-listener/blob/main/variables.tf#L81

File: README.tfdoc.hcl [hunk 13]
```diff
@@ -0,0 +1,1265 @@
...
+              User info endpoint of the IdP.
+            END
+          }
+        }
+
+        attribute "fixed_response" {
+          type        = object(fixed_response)
+          description = <<-END
+            Information for creating an action that returns a custom HTTP
+            response. Required if `type` is `fixed-response`.
```
Comment (soerenmartius) [inline README.tfdoc.hcl:359]:
```suggestion
            response.
```

File: README.tfdoc.hcl [hunk 14]
```diff
@@ -0,0 +1,1265 @@
...
+          }
+        }
+
+        attribute "authenticate_oidc" {
+          type        = object(authenticate_oidc)
+          default     = []
+          description = <<-END
+            Configuration block for an identity provider that is compliant with
+            OpenID Connect (OIDC). Specify only when `type` is
+            `authenticate-oidc`.
```
Comment (soerenmartius) [inline README.tfdoc.hcl:266]:
```suggestion
            OpenID Connect (OIDC).
```

File: README.tfdoc.hcl [hunk 15]
```diff
@@ -0,0 +1,1265 @@
...
+        }
+
+        attribute "forward" {
+          type        = object(forward)
+          description = <<-END
+            Configuration block for creating an action that distributes requests
+            among one or more target groups. Specify only if `type` is `forward`.
+            If you specify both `forward` block and `target_group_arn`
+            attribute, you can specify only one target group using `forward` and
+            it must be the same target group specified in `target_group_arn`.
```
Comment (soerenmartius) [inline README.tfdoc.hcl:393]:
```suggestion
            Configuration block for creating an action that distributes requests
            among one or more target groups.
            If you specify both `forward` block and `target_group_arn`
            attribute, you can specify only one target group using `forward` and
            it must be the same target group specified in `target_group_arn`.
```

File: README.tfdoc.hcl [hunk 16]
```diff
@@ -0,0 +1,1265 @@
...
+              END
+            }
+          }
+        }
+
+        attribute "redirect" {
+          type        = object(redirect)
+          description = <<-END
+            Configuration block for creating a redirect action. Required if
+            `type` is `redirect`.
```
Comment (soerenmartius) [inline README.tfdoc.hcl:448]:
```suggestion
            Configuration block for creating a redirect action.
```

File: README.tfdoc.hcl [hunk 17]
```diff
@@ -0,0 +1,1265 @@
...
+        attribute "id" {
+          required    = true
+          type        = string
+          description = <<-END
+            A unique identifier for the rule
+          END
+        }
+
+        attribute "priority" {
+          required    = true
```
Comment (soerenmartius) [inline README.tfdoc.hcl:534]:
```suggestion
```

Comment (soerenmartius) [reply README.tfdoc.hcl:534]: This is optional as shown in https://github.com/mineiros-io/terraform-aws-lb-listener/blob/main/rules.tf#L11

File: README.tfdoc.hcl [hunk 18]
```diff
@@ -0,0 +1,1265 @@
...
+          END
+
+          attribute "type" {
+            required    = true
+            type        = string
+            description = <<-END
+              Type of routing action. Valid values are `forward`, `redirect`,
+              `fixed-response`, `authenticate-cognito` and `authenticate-oidc`.
+            END
+          }
```
Comment (soerenmartius) [inline README.tfdoc.hcl:565]:
```suggestion
```

This has been abstracted by the sub module in https://github.com/mineiros-io/terraform-aws-lb-listener/blob/main/modules/terraform-aws-lb-listener-rule/main.tf

File: README.tfdoc.hcl [hunk 19]
```diff
@@ -0,0 +1,1265 @@
...
+          type     = object(action)
+          default = {
+            fixed_response = {
+              content_type = "plain/text"
+              message_body = "Nothing to see here!"
+              status_code  = 418
+            }
+          }
+          description = <<-END
+            Configuration block for default actions.
```
Comment (soerenmartius) [inline README.tfdoc.hcl:555]:
```suggestion
            Configuration block for the action of the rule.
```

File: README.tfdoc.hcl [hunk 20]
```diff
@@ -0,0 +1,1265 @@
...
+          }
+
+          attribute "forward" {
+            type        = object(forward)
+            description = <<-END
+              Configuration block for creating an action that distributes requests
+              among one or more target groups. Specify only if `type` is `forward`.
+              If you specify both `forward` block and `target_group_arn`
+              attribute, you can specify only one target group using `forward` and
+              it must be the same target group specified in `target_group_arn`.
```
Comment (soerenmartius) [inline README.tfdoc.hcl:614]:
```suggestion
              Configuration block for creating an action that distributes requests
              among one or more target groups.
              If you specify both `forward` block and `target_group_arn`
              attribute, you can specify only one target group using `forward` and
              it must be the same target group specified in `target_group_arn`.
```

File: README.tfdoc.hcl [hunk 21]
```diff
@@ -0,0 +1,1265 @@
...
+            required    = true
+            type        = string
+            description = <<-END
+              Domain prefix or fully-qualified domain name of the Cognito user
+              pool.
+            END
+          }
+
+          attribute "authentication_request_extra_params" {
+            type        = map(string)
```
Comment (soerenmartius) [inline README.tfdoc.hcl:909]:
```suggestion
            type        = list(authentication_request_extra_param)
```

This is a list of objects, please add the attributes of the object also

File: README.tfdoc.hcl [hunk 22]
```diff
@@ -0,0 +1,1265 @@
...
+            }
+          }
+        }
+
+        attribute "authenticate_cognito" {
+          type        = object(authenticate_cognito)
+          default     = []
+          description = <<-END
+            Configuration block for using Amazon Cognito to authenticate users.
+            Specify only when `type` is `authenticate-cognito`.
```
Comment (soerenmartius) [inline README.tfdoc.hcl:880]:
```suggestion
            Configuration block for using Amazon Cognito to authenticate users.
```

File: README.tfdoc.hcl [hunk 23]
```diff
@@ -0,0 +1,1265 @@
...
+          }
+        }
+
+        attribute "authenticate_oidc" {
+          type        = object(authenticate_oidc)
+          default     = []
+          description = <<-END
+            Configuration block for an identity provider that is compliant with
+            OpenID Connect (OIDC). Specify only when `type` is
+            `authenticate-oidc`.
```
Comment (soerenmartius) [inline README.tfdoc.hcl:952]:
```suggestion
            Configuration block for an identity provider that is compliant with
            OpenID Connect (OIDC).
```

File: README.tfdoc.hcl [hunk 24]
```diff
@@ -0,0 +1,1265 @@
...
+          attribute "issuer" {
+            required    = true
+            type        = string
+            description = <<-END
+              OIDC issuer identifier of the IdP.
+            END
+          }
+
+          attribute "authentication_request_extra_params" {
+            type        = map(string)
```
Comment (soerenmartius) [inline README.tfdoc.hcl:988]:
```suggestion
            type        = list(authentication_request_extra_param)
```

This is a list (or a set) of objects, please add the attributes also. For details please see https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lb_listener_rule

**Review by soerenmartius — CHANGES_REQUESTED** [review]

File: README.tfdoc.hcl (hunk 21 above)
Comment (kassianh) [reply README.tfdoc.hcl:909]: The schema says it is:
```
"authentication_request_extra_params": {
                                                        "type": [
                                                            "map",
                                                            "string"
                                                        ],
                                                        "description_kind": "plain",
                                                        "optional": true
                                                    },
```

I can see the terraform registry documentation differs, I'm not sure which one to go with.

**Review by mariux — APPROVED** [review]
lgtm

**Review by mariux — CHANGES_REQUESTED** [review]
needs fix but has no priority atm ;)

File: README.tfdoc.hcl (hunk 21 above)
Comment (mariux) [reply README.tfdoc.hcl:887]: let's go with provider schema

**Review by mariux — APPROVED** [review]
lgtm

**Pull request merged**